### ISSUE TITLE
Overuse of 'auto' keyword

### ISSUE THREAD

**Issue by poldon** [isocpp/CppCoreGuidelines#1881]
I realize the benefits of the **auto** keyword (templates, etc.), but I've seen multiple projects overuse it to the point where it becomes hard to follow code and determine variable types.  It not only makes it difficult to extend existing code, but it also makes C++ start to feel like a loosely-typed language.

Maybe this is a controversial topic, but I would at least like to see some guidelines on limiting the use of **auto** to cases where it's truly needed and not just another way for the developer to be lazy.

**Comment by cubbimew** [comment]
There is [ES.11: Use auto to avoid redundant repetition of type names](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-auto).

It is indeed controversial: there are guidelines in the wild ranging from AAA (almost always auto) to AAAA (almost always avoid auto).

**Comment by markgalassi** [comment]
I'm in the AAAA camp and I agree with poldon: it would be good to limit that.  The fact that many people believe in AAA is not really relevant: many people use exception for programming errors, and these guidelines take a strong stance against that (and I'm glad they do).

**Comment by markyin** [comment]
so how to do?

**Comment by neithernut** [comment]
There's [T.12](CppCoreGuidelines.md#t12-prefer-concept-names-over-auto-for-local-variables) already (though it currently looks more like a stub). Maybe that one can be extended? It's in the "Templates and genericv programming" section but it already, kind of, tells you to use concepts rather than `auto`.

----

> The fact that many people believe in AAA is not really relevant: many people use exception for programming errors, and these guidelines take a strong stance against that (and I'm glad they do).

The thing is, AAA isn't that much of an invalid standpoint. There are arguments against AAAA as well: if you need to manually keep track the type of each individual variable, i.e. you can't just trust in the interface exposed by the object, than your types are probably poorly designed, or you have a problem with spaghetti-code.

There may be valid points for keeping types, or at least their interfaces, visible in function bodies. But up until now the discussion feels more like a discussion about preferences.

**Comment by BenBE** [comment]
Maybe put some clarification on the use of `auto` urging to write code in a way where `auto` shall only be used when this doesn't cause easily recoverable loss of type information or other details about the program. In particular use of `auto` should be avoided in public interfaces unless it's use avoids obvious redundancy or dramatically reduces the type complexity of the declaration. Easily recoverable type information may e.g. be situations where you're accessing an iterator of some templated type and through the interface of the container it's obvious that you are performing such an operation.

**Comment by jwakely** [comment]
I think this kind of request for proscriptive rules about every aspect of the language is making the Core Guidelines sink under their own weight. Using `auto` or not does not cause significant bugs or maintenance problems. There are more important things to worry about.

**Comment by jwakely** [comment]
> I've seen multiple projects overuse it to the point where it becomes hard to follow code and determine variable types.

Don't do that then. If it makes the code harder to follow, it's not helping. Why do you need a rule to say code should be readable?

> it also makes C++ start to feel like a loosely-typed language.

This part is nonsense. If it feels that way to you, you don't understand how C++ works. The type system is not affected by using `auto`. Removing redundancy and verbosity is not the same as loosely-typed.

If `auto length = str.length();` isn't clear, the problem is not because of `auto`. If `auto x = var.foo();` isn't clear, it's because of bad names, not `auto`. Explicitly naming the type isn't going to help.

**Comment by poldon** [comment]
Here are a couple of examples of auto that I've seen that make code harder to follow:

1. Using auto for objects returned by functions.  (e.g. `auto obj = return_some_object();` )
2. Using auto during initialization (e.g. `auto num = 23;` )

The problem I see with Example 1 is that you don't know what object type the variable is unless you do some digging.  Granted, some IDE's may be able to handle it, but in my experience, you usually have to search the code for that function to see what it actually returns, and therefore what that object's members and functions are.

The problem with Example 2 is that the intent of the code isn't clear.  Should that be a signed int, unsigned, or 16-bit vs 32-bit, etc.  Later on, is that variable too big to fit in some other variable?  In networking code, this is critical, and auto actually makes it harder to ensure bugs don't creep in because of overflows, etc.

**Comment by rpopescu** [comment]
> The problem with Example 2 is that the intent of the code isn't clear. Should that be a signed int, unsigned, or 16-bit vs 32-bit, etc. Later on, is that variable too big to fit in some other variable? In networking code, this is critical, and auto actually makes it harder to ensure bugs don't creep in because of overflows, etc.

No, it doesn't. If one doesn't know what the type of num should be by now, they probably have no business writing this sort of code. And with respect exactly to parsing and binary manipulation, auto actually prevents many bugs, and makes refactoring easier and safer in many cases.

**Comment by neithernut** [comment]
> The problem I see with Example 1 is that you don't know what object type the variable is unless you do some digging. Granted, some IDE's may be able to handle it, but in my experience, you usually have to search the code for that function to see what it actually returns, and therefore what that object's members and functions are.

I'd say this is clearly an instance of badly designed interfaces. First of all, the type returned, or at least its relevant concept(s), should be apparent from the function declaration. I doubt anyone is arguing in favor of omitting these in _interfaces_ except some darn trivial cases and some pre-concepts TMP waiting to be rewritten or abandoned. It's not even possible if the declaration is not the definition, e.g. if you only see its declaration in a header (yes, modules...).

So yes, function and method return types should probably be reasonably exposed and not "hidden" behind an `auto`, at least not in most cases. Obviously, the function's name should also be somewhat descriptive, which should give you a hint about what the return value might be.

But even if you duplicate the function's return type verbatim in an assignment statement, a dev applying changes still needs to look up the type's or conceps' definition/documentation in order to figure out what operations it provides (assuming that the dev is not already intimate with the them). And such a dev will probably also peek at the documentation and/or declaration of the function returning that object anyway. As would the original author.

Again, nobody's going to prevent you from narrowing down the return type in a statement. To the contrary, there's [T.12](https://github.com/isocpp/CppCoreGuidelines/blob/5f393d53afcdb0f9bf7c257e4fad4aa3257753ba/CppCoreGuidelines.md#Rt-auto), already. However, arguing that `auto` would introduce errors sounds more like treating the symptoms of badly designed types and interfaces. If that were the case then the same would probably also apply to temporaries.

**Comment by hsutter** [comment]
Editors call: We don't think there's anything to add at this point over the current guidelines. We can't legislate against the overuse of every feature.
